jax/experimental/mosaic/gpu/utils.py

```diff
@@ -1365,3 +1365,80 @@ def vector_concat(vectors: Sequence[ir.Value]) -> ir.Value:
       result = vector.insertelement(elem, result, position=c(offset + i, index))
     offset += vty.shape[0]
   return result
+
+
+def signal_with_red(mc_ptr, is_relaxed=False):
+  mode = "relaxed" if is_relaxed else "release"
+  asm_instr = f"""
+  {{
+      multimem.red.{mode}.sys.global.add.u32 [$0], 1;
+      fence.proxy.alias;
+  }}"""
+  llvm.inline_asm(
+    ir.Type.parse("!llvm.void"),
+    [mc_ptr],
+    asm_instr,
+    "l",
+    has_side_effects=True,
+    asm_dialect=0,
+  )
+
+
+def wait_loop(uc_ptr, num_gpus=8, is_relaxed=False):
+  mode = "relaxed" if is_relaxed else "acquire"
+  asm_instr = f"""
+  {{
+      .reg .u32   %tmp32_<1>;
+      .reg .pred  %p<1>;
+
+      wait_signal:
+          atom.global.sys.{mode}.cas.b32 %tmp32_0, [$0], {num_gpus}, 0;
+          setp.eq.u32 %p0, %tmp32_0, 8;
+          @!%p0 bra wait_signal;
+  }}"""
+  llvm.inline_asm(
+    ir.Type.parse("!llvm.void"),
+    [uc_ptr],
+    asm_instr,
+    "l",
+    has_side_effects=True,
+    asm_dialect=0,
+  )
+
+
+def global_membar():
+  llvm.inline_asm(
+      ir.Type.parse("!llvm.void"),
+      [],
+      "membar.gl;",
+      "",
+      has_side_effects=True,
+  )
+
+
+def multimem_ld_reduce_128(mc_ptr):
+  i32 = ir.IntegerType.get_signless(32)
+  return_struct = llvm.inline_asm(
+      ir.Type.parse("!llvm.struct<(i32,i32,i32,i32)>"),
+      [mc_ptr],
+      "multimem.ld_reduce.relaxed.sys.global.add.v4.f16x2 {$0, $1, $2, $3}, [$4];",
+      "=r,=r,=r,=r,l",
+      has_side_effects=True,
+      asm_dialect=0,
+  )
+  return_regs = [
+        llvm.extractvalue(i32, return_struct, [i]) for i in range(4)
+  ]
+  return return_regs[0], return_regs[1], return_regs[2], return_regs[3]
+
+
+def multimem_st_128(mc_ptr, x, y, z, w):
+  i32 = ir.IntegerType.get_signless(32)
+  llvm.inline_asm(
+      i32,
```

Comment:
This snippet does not have a result. Use `ir.Type.parse("!llvm.void")` and remove the result register constraint